File: practice_programming/0003. Universal Travel Sites/max_flow.cpp
/*
Sure, let's implement the solution in C++ using the Edmonds-Karp algorithm to find the maximum flow in a flow network.

### Implementation in C++

Here's the full implementation:


### Explanation:

1. **Reading Input**:
   - Read the source and destination planets, and the number of connections.
   - Use an `unordered_map` to map planet names to unique indices.
   - Store the edges (connections) in a `vector` of tuples.

2. **Graph Construction**:
   - Construct an adjacency matrix `graph` to represent the capacities between planets.
   - Convert planet names to their corresponding indices using the `unordered_map`.

3. **BFS Function**:
   - Perform BFS to find an augmenting path in the residual graph.
   - Update the `parent` vector to record the path.

4. **Edmonds-Karp Algorithm**:
   - Initialize the residual graph.
   - Use BFS to find augmenting paths and update the residual capacities.
   - Accumulate the maximum flow.

5. **Main Function**:
   - Read the input, build the graph, and call the `edmondsKarp` function.
   - Print the result.

### Running the Script:
To compile and run the script, save it as `max_flow.cpp` and use the following commands:

```sh
g++ -o max_flow max_flow.cpp
./max_flow < input.txt
```

Or you can provide the input directly in the terminal using echo and pipe:

```sh
echo "EAR MAR 11 EAR AAA 300 EAR BBB 400 AAA BBB 100 AAA CCC 400 AAA MAR 300 BBB DDD 400 AAA DDD 400 DDD AAA 100 CCC MAR 400 DDD CCC 200 DDD MAR 300" | ./max_flow
```

This will read the input, build the graph, compute the maximum flow, and print the minimum capacity required for the planet station.
*/

#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <string>
#include <unordered_map>

using namespace std;

// Function to perform BFS and find if there is a path from source to sink
bool bfs(vector<vector<int>>& residualGraph, int source, int sink, vector<int>& parent) {
    parent.clear();
    
    int V = residualGraph.size();
    vector<bool> visited(V, false);
    queue<int> q;
    q.push(source);
    visited[source] = true;
    parent[source] = -1;

    while (!q.empty()) {
        int u = q.front();
        q.pop();

        for (int v = 0; v < V; v++) {
            if (!visited[v] && residualGraph[u][v] > 0) {
                q.push(v);
                parent[v] = u;
                visited[v] = true;
                if (v == sink) {
                    return true;
                }
            }
        }
    }

    return false;
}

// Implementation of Edmonds-Karp algorithm for finding the maximum flow
int edmondsKarp(vector<vector<int>>& graph, int source, int sink) {
    int V = graph.size();
    vector<vector<int>> residualGraph(V, vector<int>(V));
    for (int u = 0; u < V; u++) {
        for (int v = 0; v < V; v++) {
            residualGraph[u][v] = graph[u][v];
        }
    }

    vector<int> parent(V);
    int maxFlow = 0;

    while (bfs(residualGraph, source, sink, parent)) {
        int pathFlow = INT_MAX;
        for (int v = sink; v != source; v = parent[v]) {
            int u = parent[v];
            pathFlow = min(pathFlow, residualGraph[u][v]);
        }

        for (int v = sink; v != source; v = parent[v]) {
            int u = parent[v];
            residualGraph[u][v] -= pathFlow;
            residualGraph[v][u] += pathFlow;
        }

        maxFlow += pathFlow;
    }

    return maxFlow;
}

int main() {
    string sourcePlanet, destinationPlanet;
    int N;
    cin >> sourcePlanet >> destinationPlanet >> N;

    // Mapping planet names to unique indices
    unordered_map<string, int> planetIndex;
    vector<tuple<string, string, int>> edges;

    int index = 0;
    for (int i = 0; i < N; i++) {
        string src, dest;
        int capacity;
        cin >> src >> dest >> capacity;

        if (planetIndex.find(src) == planetIndex.end()) {
            planetIndex[src] = index++;
        }
        if (planetIndex.find(dest) == planetIndex.end()) {
            planetIndex[dest] = index++;
        }

        edges.push_back(make_tuple(src, dest, capacity));
    }

    int V = planetIndex.size();
    // adjacency matrix to represent the capacities between planets
    vector<vector<int>> graph(V, vector<int>(V, 0));

    for (auto& [src, dest, capacity] : edges) {
        int u = planetIndex[src];
        int v = planetIndex[dest];
        graph[u][v] = capacity;
    }

    int source = planetIndex[sourcePlanet];
    int sink = planetIndex[destinationPlanet];

    int maxFlow = edmondsKarp(graph, source, sink);
    cout << maxFlow << endl;

    return 0;
}

File: practice_programming/0003. Universal Travel Sites/max_flow.py
"""
keyword: graph, flow network, Ford-Fulkerson, Edmonds-Karp, BFS, finding the maximum flow in a flow network
https://en.wikipedia.org/wiki/Flow_network
https://en.wikipedia.org/wiki/Ford%E2%80%93Fulkerson_algorithm
https://en.wikipedia.org/wiki/Edmonds%E2%80%93Karp_algorithm

To solve this problem, we need to find the minimum capacity of a planet station that allows every space vessel to dock and download its passengers on arrival, ensuring that the source planet can reach the destination planet. This is a classic example of finding the maximum flow in a flow network, which can be solved using algorithms like the Ford-Fulkerson method or Edmonds-Karp algorithm (a specific implementation of the Ford-Fulkerson method using BFS).

Here’s how to approach this problem using Python:

1. **Model the problem as a flow network**:
   - Each planet station is a node.
   - Each connection between stations is an edge with a capacity (maximum number of passengers).

2. **Use the Edmonds-Karp algorithm to find the maximum flow**:
   - Initialize a residual graph.
   - Use BFS to find augmenting paths.
   - Update the residual capacities of the edges and reverse edges.

### Implementation in Python


### Explanation:
1. **Reading Input**:
   - The input is read from standard input and split into a list of strings.
   - The source and destination planets and the number of connections are extracted.
   - The connections are read into a graph represented by a `defaultdict` of `defaultdicts` to store capacities.

2. **BFS Function**:
   - This function checks for an augmenting path in the residual graph and updates the parent map to record the path.

3. **Edmonds-Karp Algorithm**:
   - Initializes the residual graph.
   - Uses BFS to find augmenting paths and updates the residual capacities.
   - Accumulates the maximum flow.

4. **Main Function**:
   - Parses the input and builds the graph.
   - Calls `edmonds_karp` to find the maximum flow from the source to the destination.
   - Prints the result.

### Running the Script:
To run this script, save it as `max_flow.py` and provide the input through stdin:

```sh
python max_flow.py < input.txt
```

Or you can provide the input directly in the terminal using echo and pipe:

```sh
echo "EAR MAR 11 EAR AAA 300 EAR BBB 400 AAA BBB 100 AAA CCC 400 AAA MAR 300 BBB DDD 400 AAA DDD 400 DDD AAA 100 CCC MAR 400 DDD CCC 200 DDD MAR 300" | python max_flow.py
```

This will read the input, build the graph, compute the maximum flow, and print the minimum capacity required for the planet station.

"""


from collections import deque, defaultdict
from pprint import pprint
from typing import Dict

# Function to perform BFS and find if there is a path from source to sink
def bfs(residual_graph: Dict[str, Dict[str, int]], source: str, sink: str, parent: Dict[str, str]) -> bool:
    parent.clear()
    visited = set()
    queue = deque([source])
    visited.add(source)
    
    while queue:
        u = queue.popleft()
        
        for v, capacity in residual_graph[u].items():
            if v not in visited and capacity > 0:
                queue.append(v)
                visited.add(v)
                parent[v] = u
                if v == sink:
                    return True
    return False

# Implementation of Edmonds-Karp algorithm for finding the maximum flow
def edmonds_karp(graph: Dict[str, Dict[str, int]], source: str, sink: str) -> int:
    """
    The reverse edges are crucial for two reasons:
    1. Flow Adjustment: If in a future iteration, we find that pushing flow through a different path requires reducing the flow along a previously used path, the reverse edges allow us to do so.
    2. Correctness: They ensure the algorithm can correctly backtrack and reallocate flow to achieve the maximum flow in the network.

    Args:
        graph (Dict[str, Dict[str, int]]): _description_
        source (str): _description_
        sink (str): _description_

    Returns:
        _type_: _description_
    """
    residual_graph = defaultdict(lambda: defaultdict(int))
    
    for u in graph:
        for v, capacity in graph[u].items():
            residual_graph[u][v] = capacity
    
    # parent[s] = the parent of node s in the path from source to sink, used to trace the path from sink to source
    parent :Dict[str, str]= {}
    max_flow = 0
    
    while bfs(residual_graph, source, sink, parent):
        path_flow = float('Inf')
        s = sink
        
        while s != source:
            path_flow = min(path_flow, residual_graph[parent[s]][s])
            s = parent[s]
        
        v = sink
        while v != source:
            u = parent[v]
            residual_graph[u][v] -= path_flow
            #  These reverse edges represent the possibility to reduce the flow along an edge.
            residual_graph[v][u] += path_flow
            v = u
        
        max_flow += path_flow
    
    return max_flow

def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    
    source_planet = data[0]
    destination_planet = data[1]
    N = int(data[2])
    
    #  adjacency lists to represent the graph
    graph = defaultdict(lambda: defaultdict(int))
    
    index = 3
    for _ in range(N):
        src = data[index]
        dest = data[index + 1]
        capacity = int(data[index + 2])
        graph[src][dest] += capacity
        index += 3
    
    # print(graph)
    # pprint(graph)

    # Calculate the maximum flow
    max_capacity = edmonds_karp(graph, source_planet, destination_planet)
    
    # Output the result
    print(max_capacity)

if __name__ == "__main__":
    main()

